mods/weather/init.lua

```diff
@@ -114,22 +103,93 @@ local function update_clouds()
 			speed = {x = n_speedx * 4, z = n_speedz * 4},
 		})
 		-- now adjust the shadow intensity
-		player:set_lighting({ shadows = { intensity = 0.7 * (1 - density) } })
+		player:set_lighting({
+			shadows = { intensity = 0.7 * (1 - density) }
+		})
 	end
 end
 
+local function purge_effects(player)
+	-- reset potentially touched values to their defaults
+	if randomize_clouds then
+		player:set_clouds({
+			density = 0.4,
+			thickness = 16,
+			speed = { x = 0, z = -2 }
+		})
+	end
+	player:set_lighting({
+		shadows = { intensity = 0 }
+	})
+end
 
-local function cyclic_update()
-	update_clouds()
-	minetest.after(CYCLE, cyclic_update)
+-- Define API hooks
+
+-- override to set state for newly joined players
+weather.enable_on_join = true
+
+-- returns bool for whether weather is active
+-- returns nil if player is offline or weather is disabled
+weather.get_enabled = function(player)
+	return playerlist[player]
 end
 
+-- override weather generation for individual player
+weather.set_enabled = function(player, enable)
+	if enable == playerlist[player] then
+		return
+	end
+	playerlist[player] = enable
+	if enable then
+		if randomize_clouds then
+			update_clouds({player})
+		end
+	else
+		purge_effects(player)
+	end
+end
 
-minetest.after(0, cyclic_update)
+-- End API hooks
 
+-- skip event registration if mod is disabled
+if not mod_enabled then
+	return
+end
+
+if update_clouds then
+	local function cyclic_update()
+		local players = {}
+		for player, state in pairs(playerlist) do
+			if state then
+				table.insert(players, player)
+			end
+		end
+		update_clouds(players)
```

Comment:
This would change to `update_clouds(playerset)`; the for-loop above would be obsolete.